File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/PercentageRange.java
package pl.allegro.experiments.chi.chiserver.domain.experiments;

import com.google.common.base.Preconditions;

public final class PercentageRange extends Range {
    public PercentageRange(int from, int to) {
        super(from, to);
        Preconditions.checkArgument(to <= 100, "Range.to > 100");
    }

    public int size() {
        return getTo() - getFrom();
    }

    @Override
    public boolean equals(Object o) {
        return super.equals(o);
    }

    public PercentageRange trim(int newSize, boolean fromRight) {
        Preconditions.checkArgument(newSize <= size());

        if (fromRight) {
            return new PercentageRange(getTo() - newSize, getTo());
        } else {
            return new PercentageRange(getFrom(), getFrom() + newSize);
        }
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/PercentageRange.kt
package pl.allegro.experiments.chi.chiserver.domain

import com.google.common.base.Preconditions

class PercentageRange(from: Int, to: Int) : Range<Int>(from, to) {
    init {
        Preconditions.checkArgument(to <= 100, "Range.to > 100 in PercentageRange($from..$to)")
    }

    override operator fun contains(value: Int): Boolean {
        return value in from..(to - 1)
    }
}
